contracts/registrar/registrar.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package registrar
+
+//go:generate abigen --sol contract/registrar.sol --pkg contract --out contract/registrar.go
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/registrar/contract"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type Registrar struct {
+	contract *contract.Contract
+}
+
+// NewRegistrar binds checkpoint contract and returns a registrar instance.
+func NewRegistrar(contractAddr common.Address, backend bind.ContractBackend) (*Registrar, error) {
+	c, err := contract.NewContract(contractAddr, backend)
+	if err != nil {
+		return nil, err
+	}
+	return &Registrar{contract: c}, nil
+}
+
+// Contract returns the underlying contract instance.
+func (registrar *Registrar) Contract() *contract.Contract {
+	return registrar.contract
+}
+
+// LookupCheckpointEvent searches checkpoint event for specific section in the
+// given log batches.
+func (registrar *Registrar) LookupCheckpointEvent(blockLogs [][]*types.Log, section uint64, hash common.Hash) []*contract.ContractNewCheckpointVote {
+	var votes []*contract.ContractNewCheckpointVote
+
+	for _, logs := range blockLogs {
+		for _, log := range logs {
+			event, err := registrar.contract.ParseNewCheckpointVote(*log)
+			if err != nil {
+				continue
+			}
+			if event.Index == section && common.Hash(event.CheckpointHash) == hash {
+				votes = append(votes, event)
+			}
+		}
+	}
+	return votes
+}
+
+// RegisterCheckpoint registers the checkpoint with a batch of associated signatures
+// that are collected off-chain and sorted by lexicographical order.
```

Comment:
Should the sorting be done outside of this function -- and if so, is that a manual thing? Would be nice if the interface sorts it for us, if so

Reply:
I put the sorting work in the `cmd/registrar/exec.go`. Since we need to sort the signature by "signer address", if we put the sorting logic here we also need to pass the signer list.
But both inside or outside it's fine for me. Anyway we don't need to sort them manually, `registrar` should do everything for us.

Reply:
:+1: